File: Platform/Workspace/Typescript/framework/data/Sort.ts
import { RoleType } from '../meta';

export interface SortArgs {
  roleType: RoleType;
  descending?: boolean;
}

export class Sort {
  public roleType: RoleType;
  public descending: boolean;

  constructor(args: SortArgs);
  constructor(roleType: RoleType, descending?: boolean);
  constructor(args: SortArgs | RoleType, descending?: boolean) {
    if (args instanceof RoleType) {
      this.roleType = args;
      this.descending = descending ?? false;
    } else if (args) {
      this.roleType = args.roleType;
      this.descending = args.descending ?? false;
    }
  }

  public toJSON(): any {
    return {
      roleType: this.roleType.id,
      descending: this.descending,
    };
  }
}

File: Platform/Database/Allors.Database/Data/Sort.cs
// <copyright file="Sort.cs" company="Allors bv">
// Copyright (c) Allors bv. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Data
{
    using Allors.Meta;

    public class Sort
    {
        public Sort(IRoleType roleType = null) => this.RoleType = roleType;

        public IRoleType RoleType { get; set; }

        public bool Descending { get; set; }

        public void Build(Allors.Extent extent) => extent.AddSort(this.RoleType, this.Descending ? SortDirection.Descending : SortDirection.Ascending);
    }
}
